### PR TITLE
feat: snapmirror, volume plugin

### PR DESCRIPTION
**Pull request by Hardikl**
image

image

image

image

### CODE REVIEW

**Comment by Hardikl** [comment]
```
hardikl@hardikl-mac-0 harvest % bin/harvest doctor diffzapirest --zapidatacenter zapi --restdatacenter rest
################## Label value Diffs in prometheus ##############
******* Mismatch in label length. Check data volume_labels aggr
******* Mismatch in label length. Check data volume_labels node
******* Mismatch in label length. Check data volume_labels snapshot_policy
******* Mismatch in label length. Check data volume_labels all_sm_healthy
******* Mismatch in label length. Check data node_labels location
******* Mismatch in label length. Check data qtree_labels qtree
******* Mismatch in label length. Check data snapmirror_labels source_node
******* Mismatch in label length. Check data snapmirror_labels last_transfer_type
******* Mismatch in label length. Check data snapmirror_labels schedule
******* Mismatch in label length. Check data snapmirror_labels unhealthy_reason
################## Missing Labels ##############
shelf_fan_labels []
shelf_voltage_labels []
shelf_labels []
disk_labels []
lun_labels []
qtree_labels []
shelf_sensor_labels []
snapmirror_labels []
volume_labels []
aggr_labels []
node_labels []
shelf_psu_labels [location]
shelf_temperature_labels []
################## Metrics Diffs in prometheus ##############
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
strconv.ParseFloat: parsing "": invalid syntax
################## Missing Metrics by Object in prometheus ##############
qtree [qtree_threshold]
cluster [cluster_outstanding_alerts cluster_suppressed_alerts]
aggr [aggr_space_performance_tier_inactive_user_data aggr_space_reserved aggr_inode_maxfiles_custom_possible aggr_space_performance_tier_inactive_user_data_percent aggr_aggr_inode_maxfiles_custom_possible aggr_space_physical_used_percent aggr_snapshot_inode_used_percent aggr_hybrid_cache_size_total aggr_maxfiles_custom_possible]
################## Missing Metrics from dashboard ##############
aggr [aggr_snapshot_inode_used_percent]
##################
hardikl@hardikl-mac-0 harvest % 
```

File: cmd/collectors/rest/plugins/snapmirror/snapmirror.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		nodeName := volume.Get("node").String()
+		my.svmVolToNode[vserverName+volumeName] = nodeName
+	}
+	return nil
+}
+
+func (my *SnapMirror) updateProtectedFields(instance *matrix.Instance) {
+
+	// check for group_type
+	if instance.GetLabel("group_type") != "" {
```
Comment (cgrinds) [inline cmd/collectors/rest/plugins/snapmirror/snapmirror.go:123]: this block looks pretty similar to a block in volume.go - can each plugin share this same block?

File: cmd/collectors/rest/rest.go [hunk 2]
```diff
@@ -372,7 +374,8 @@ func (r *Rest) PollData() (*matrix.Matrix, error) {
 			f := instanceData.Get(metric.name)
 			if f.Exists() {
 				metr.SetName(metric.label)
-				if err = metr.SetValueFloat64(instance, f.Float()); err != nil {
+				floatValue := HandleIfTimeField(f.String())
```
Comment (cgrinds) [inline cmd/collectors/rest/rest.go]: as discussed in chat, let's try an explicit solution

Comment (Hardikl) [reply cmd/collectors/rest/rest.go]: Implemented counter(type) way.
May be I have submitted latest change while you were reviewing previous version of PR :)

**Comment by cgrinds** [comment]
we also need to update the template DSL documentation with the new metrics type notation

File: cmd/collectors/rest/plugins/volume/volume.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+	my.query = "api/private/cli/snapmirror"
+	my.data = matrix.New(my.Parent+".Volume", "volume", "volume")
+
+	my.outgoingSM = make(map[string][]string)
+	my.incomingSM = make(map[string]string)
+	my.isHealthySM = make(map[string]bool)
+
+	// Assigned the value to currentVal so that plugin would be invoked first time to populate cache.
+	if my.currentVal, err = my.setPluginInterval(); err != nil {
+		my.Logger.Error().Err(err).Stack().Msg("Failed while setting the plugin interval")
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/volume/volume.go:73]: it should return err to caller

File: cmd/collectors/rest/plugins/volume/volume.go [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+	// Set all global labels from zapi.go if already not exist
+	my.data.SetGlobalLabels(data.GetGlobalLabels())
+
+	if my.currentVal >= my.pluginInvocationRate {
+		my.currentVal = 0
+
+		// invoke snapmirror zapi and populate info in source and destination snapmirror maps
+		smSourceMap, smDestinationMap, err := my.GetSnapMirrors()
+		if err != nil {
+			my.Logger.Error().Stack().Err(err).Msg("Failed to collect snapmirror data")
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/volume/volume.go:94]: nope, both smSourceMap and smDestinationMap would be nil in that case. line 98 function call moved to else block.

File: cmd/collectors/rest/plugins/volume/volume.go [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+	"goharvest2/pkg/dict"
+	"goharvest2/pkg/errors"
+	"goharvest2/pkg/matrix"
+	"goharvest2/pkg/tree/node"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const DefaultPluginInvocationRate = 10
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/volume/volume.go:22]: Agreed, made the changes for duration where time is used.

File: cmd/collectors/rest/plugins/volume/volume.go [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+	var dataIntervalStr = ""
+	schedule := param.GetChildS("schedule")
+	if schedule != nil {
+		dataInterval := schedule.GetChildS("data")
+		if dataInterval != nil {
+			dataIntervalStr = dataInterval.GetContentS()
+		}
+	}
+
+	// Convert the interval from str to int
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/volume/volume.go:128]: yes, added.

File: cmd/collectors/rest/plugins/volume/volume.go [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+			protectedByMap[smRelationship.GetLabel("protectedBy")] = ""
+
+			// Update isHealthySM map based on the source snapmirror info
+			if volumeType == "rw" {
+				/* Example: If 3 relationships belongs to a volume, and out of 3, 2 are healthy and one is not,
+				   So, isHealthySM map value would be unhealthy - false
+				*/
+				currentVal, _ := strconv.ParseBool(smRelationship.GetLabel("is_healthy"))
+				healthStatus = healthStatus && currentVal
+				my.isHealthySM[key] = healthStatus
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/volume/volume.go:292]: Yes, `protectedByMap[k]` = "" is still needed for each iteration, where k can be volume, storage_vm, or cg. we can't break the inner loop.

File: cmd/collectors/rest/plugins/snapmirror/snapmirror.go (hunk 1 above)
Comment (Hardikl) [reply cmd/collectors/rest/plugins/snapmirror/snapmirror.go:123]: it's almost same, moved to one util class and re-used.

File: cmd/collectors/rest/rest.go [hunk 8]
```diff
@@ -372,7 +382,19 @@ func (r *Rest) PollData() (*matrix.Matrix, error) {
 			f := instanceData.Get(metric.name)
 			if f.Exists() {
 				metr.SetName(metric.label)
-				if err = metr.SetValueFloat64(instance, f.Float()); err != nil {
+				var floatValue float64
+				if metric.metricType != "" {
+					switch metric.metricType {
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/rest.go:387]: done

File: cmd/collectors/rest/rest.go [hunk 9]
```diff
@@ -175,8 +178,15 @@ func (r *Rest) initEndPoints() error {
 				}
 				if line1.GetNameS() == "counters" {
 					for _, c := range line1.GetAllChildContentS() {
+						mType := ""
 						name, display, kind = util.ParseMetric(c)
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/rest.go:182]: moved metricType inside ParseMetric function.

File: cmd/collectors/rest/rest.go [hunk 10]
```diff
@@ -475,44 +392,7 @@ func (r *Rest) processEndPoints(data *matrix.Matrix) error {
...
-					f := instanceData.Get(metric.name)
-					if f.Exists() {
-						metr.SetName(metric.label)
-						if err = metr.SetValueFloat64(instance, f.Float()); err != nil {
-							r.Logger.Error().Err(err).Str("key", metric.name).Str("metric", metric.label).
-								Msg("Unable to set float key on metric")
-						}
-					}
-				}
+				r.HandleLabelsAndMetrics(instance, endpoint.prop, instanceData, data, instanceKey)
```
Comment (rahulguptajss) [inline cmd/collectors/rest/rest.go:395]: Can you simplify method signature here. I think some of these you can get from Rest collector reference.

File: cmd/collectors/rest/templating.go [hunk 11]
```diff
@@ -58,57 +64,230 @@ func (r *Rest) initCache() error {
...
+			}
+			seconds += f
+		}
+		return seconds
+	}
+
+	return 0
+}
+
+func HandleTimestamp(value string) float64 {
```
Comment (rahulguptajss) [inline cmd/collectors/rest/templating.go:144]: looks like a method better suited for templating. Same for all other similiar methods.

File: cmd/collectors/rest/templating.go [hunk 12]
```diff
@@ -58,57 +64,230 @@ func (r *Rest) initCache() error {
...
 			}
 		}
 	}
 
-	r.Logger.Info().Strs("extracted Instance Keys", r.prop.instanceKeys).Msg("")
-	r.Logger.Info().Int("count metrics", len(r.prop.metrics)).Int("count labels", len(r.prop.instanceLabels)).Msg("initialized metric cache")
+}
 
-	return nil
+func (r *Rest) HandleLabelsAndMetrics(instance *matrix.Instance, prop *prop, instanceData gjson.Result, data *matrix.Matrix, instanceKey string) uint64 {
```
Comment (rahulguptajss) [inline cmd/collectors/rest/templating.go:212]: should be in Rest.go. Nothing to do with template parsing.

File: cmd/collectors/rest/templating.go [hunk 13]
```diff
@@ -58,57 +64,230 @@ func (r *Rest) initCache() error {
...
+				r.Logger.Error().Err(err).Str("key", metric.name).Str("metric", metric.label).
+					Msg("Unable to set float key on metric")
+			}
+			count++
+		}
+	}
+	return count
+}
+
+func (r *Rest) GetRestData(prop *prop, client *rest.Client) ([]interface{}, error) {
```
Comment (rahulguptajss) [inline cmd/collectors/rest/templating.go:273]: should be in Rest.go

File: cmd/collectors/rest/rest.go (hunk 10 above)
Comment (Hardikl) [reply cmd/collectors/rest/rest.go:395]: Agree, but function can be called from 2 places, 1 at main api and 2 at endpoint, so based on the caller, args will be changed.

File: cmd/collectors/rest/templating.go (hunk 11 above)
Comment (Hardikl) [reply cmd/collectors/rest/templating.go:144]: yes.

File: cmd/collectors/rest/templating.go (hunk 12 above)
Comment (Hardikl) [reply cmd/collectors/rest/templating.go:212]: done

File: cmd/collectors/rest/templating.go (hunk 13 above)
Comment (Hardikl) [reply cmd/collectors/rest/templating.go:273]: done

File: cmd/collectors/rest/plugins/snapmirror/snapmirror.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+	numRecords := results[0]
+	if numRecords.Int() == 0 {
+		return errors.New(errors.ERR_NO_INSTANCE, "no "+my.query+" instances on cluster")
+	}
+
+	for _, volume := range results[1].Array() {
+		volumeName := volume.Get("volume").String()
+		vserverName := volume.Get("vserver").String()
+		nodeName := volume.Get("node").String()
+		my.svmVolToNode[vserverName+volumeName] = nodeName
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/snapmirror/snapmirror.go:115]: clean map

File: cmd/collectors/rest/rest.go [hunk 15]
```diff
@@ -419,16 +344,8 @@ func (r *Rest) processEndPoints(data *matrix.Matrix) error {
...
-
-		r.Logger.Debug().Str("href", href).Msg("")
-
-		if href == "" {
-			return errors.New(errors.ERR_CONFIG, "empty url")
-		}
-		err = rest.FetchData(r.client, href, &records)
-		if err != nil {
-			r.Logger.Error().Stack().Err(err).Str("href", href).Msg("Failed to fetch data")
+		if records, err = r.GetRestData(endpoint.prop, r.client); err != nil {
```
Comment (Hardikl) [inline cmd/collectors/rest/rest.go:347]: parse time to cover endpoint

File: cmd/collectors/rest/rest.go (hunk 10 above)
Comment (Hardikl) [inline cmd/collectors/rest/rest.go:395]: refactor to cover for loop inside function

File: conf/rest/9.12.0/snapmirror.yaml [hunk 16]
```diff
@@ -0,0 +1,63 @@
+
```
Comment (Hardikl) [inline conf/rest/9.12.0/snapmirror.yaml:1]: why we use private cli?

Comment (Hardikl) [reply conf/rest/9.12.0/snapmirror.yaml:2]: added comment

File: cmd/collectors/rest/rest.go (hunk 10 above)
Comment (Hardikl) [reply cmd/collectors/rest/rest.go]: done

File: cmd/collectors/rest/rest.go (hunk 15 above)
Comment (Hardikl) [reply cmd/collectors/rest/rest.go:309]: will check the usage of parse_time in metadata.

File: cmd/collectors/rest/plugins/snapmirror/snapmirror.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/snapmirror/snapmirror.go]: done

File: cmd/collectors/rest/plugins/snapmirror/snapmirror.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+	"goharvest2/cmd/collectors/rest/plugins"
+	"goharvest2/cmd/poller/plugin"
+	"goharvest2/cmd/tools/rest"
+	"goharvest2/pkg/conf"
+	"goharvest2/pkg/errors"
+	"goharvest2/pkg/matrix"
+	"time"
+)
+
+const PluginInvocationRate = 10
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/snapmirror/snapmirror.go:18]: Here, it's not needed as we don't provide the configurable schedule for snapmirror plugin.
image

File: cmd/collectors/rest/plugins/restpluginutil.go [hunk 18]
```diff
@@ -0,0 +1,62 @@
+package plugins
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/restpluginutil.go]: done

**Review by rahulguptajss — APPROVED** [review]

**Review by rahulguptajss — APPROVED** [review]

**Pull request merged**